### DISCUSSION TITLE
Describe your changes

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#3800] [Announcements]
Everyone makes mistakes, I have most certainly made enough of them here in `rust-bitcoin`. 

When chasing bugs or doing archival work in the git index it is much more difficult if one lands on a commit with a single line commit log. Especially if it just says "Do XYZ" (i.e., adds no additional information to the code change).

I would like it propose a hard-ish policy, especially for newer devs or devs without a reputation, that every change (that means _every_ patch) has a commit log that includes a 'why' section and a 'what' section. Sure, for trivial patches the 'what' section can be left off, especially if the title is the 'what'.

In my earlier life I learned a whole ton of things because Greg KH brutally encouraged me to write why and what sections of my commits. Not only does it make review easier (primarily it answers the question "does this dev even understand what these changes are doing and have they thought about the consequences") but it also teaches newer devs a bunch of things.

1. How to do technical writing
2. How to explain _succinctly_ what they are doing
3. Makes them look up jargon and correct terms 

This is a non-trivial library, its not a great place for learning development but for those that want to do so we should encourage them but hold them to high standards and help them to become better developers.

The canonical blog post (IMO) describing how to write a commit log: https://cbea.ms/git-commit/

**Comment by apoelstra** [comment]
Sounds good. Feel free to yell at me for not doing this.

**Reply by storopoli** [reply]
+1 for me as well. Loved the idea.

**Reply by shinghim** [reply]
I like this idea. I'll try to remember this, but feel free to yell at me for this as well

**Comment by yancyribbens** [comment]
@tcharding if you want to institute new policy's on the project, lead by example, don't make this about whomever you perceive to be a new developer without reputation.  Make policy that brings everyone up, not trys to keep a few people down.

**Reply by apoelstra** [reply]
@tcharding's commit messages are not perfect (and don't have explicit "why" and "what" headings) but after quickly checking the last couple weeks of work, they're all pretty thorough and descriptive.

**Reply by yancyribbens** [reply]
I'm all for suggestions if it's about bringing everyone's code quality up.

**Reply by tcharding** [reply]
For the last few months I have personally been trying to write less because I thought I tend to write too much and I feel it abuses peoples time to expect them to read them.

FTR I never meant explicit sections with headings just a line or two that says why the change is needed followed by a line or two (separate paragraph) that says what the patch does. This is the commit log format used by the Linux kernel.

**Reply by tcharding** [reply]
Ha amusingly the kernel docs have the same heading I used in this discussion, that was not intentional.

https://www.kernel.org/doc/html/v4.10/process/submitting-patches.html#describe-your-changes

**Reply by yancyribbens** [reply]
Yeah from the 4.10 kernel :P

https://www.kernel.org/doc/html/v6.9/process/5.Posting.html

sometimes I wish we could go back to sending email patches for things.  actually most times.

**Reply by Kixunil** [reply]
> I thought I tend to write too much and I feel it abuses peoples time to expect them to read them.

From my point of view, it doesn't since being clear saves communication rounds over ridiculous timezones. Actually, now I realize that I need to improve details in my *review comments* (unless it bothers you)

**Reply by tcharding** [reply]
I like the details, please do.

**Comment by Kixunil** [comment]
Awesome! I already do this for years (I think since about 2016 or 2017) and it can be very helpful when digging into the past, especially if the code is somehow weird. However to be really helpful, *if the change is weird* the why needs to be very detailed. I'd suggest we have a rule for this. If any of the maintainers needs to ask questions to understand the change then the commit message is not good enough. (Don't worry, this should be rare.)

That being said, I believe we can safely have exceptions for these obvious changes:

* typo fixes in docs, comments or names of non-public items
* warning fixes that were **not** "fixed" by silencing
* running formatter

Also sometimes, when the code is particularly weird, it's better to write a comment in the code directly. In those cases I'd accept a commit message with descriptive title, short summary and a recommendation to read the code comment.

**Reply by tcharding** [reply]
Yep agreed, I'm not 100% sure of the meaning of point 2 but I get the overall idea.

**Reply by Kixunil** [reply]
By silencing I mean `#[allow]`. I want the reason why `#[allow]` was chosen spelled out.

**Reply by tcharding** [reply]
Ah yeah, me to. I mentioned as much on a PR this week somwhere.
